README.md
# groups-wc-not-purchasable
This plugin is an extension to Groups and WooCommerce to have products that can not be purchased by group members.

This plugin can be used to have products that can not be purchased by group members.

## Requirements
- Required : [Groups](http://wordpress.org/plugins/groups/)
- Required : [WooCommerce](http://wordpress.org/plugins/woocommerce/)
- Recommended : [Groups WooCommerce](https://woocommerce.com/products/groups-woocommerce/)

To make appropriate use of this plugin following the examples given below, you should be using [Groups WooCommerce](https://woocommerce.com/products/groups-woocommerce/).

## Usage

Usage example based on the first filter example as shown below.

- You must have [Groups](http://wordpress.org/plugins/groups/) and [WooCommerce](http://wordpress.org/plugins/woocommerce/) activated.
- Install and activate this plugin.
- Create a group named "Premium".
- Assign a user to the "Premium" group.
- Assign one or more products to a category named "Standard".
- Add the filter hook to your `functions.php` as shown below.
- Users in the "Premium" group can not purchase the products in the "Standard" category.
- Any visitor is allowed to see those products, but members of the excluded groups cannot purchase them.

## Using a filter hook to set up which product categories cannot be purchased by which groups.

We will use the filter `groups_wc_not_purchasable_category_to_group` to determine relationships between product categories and groups - group members will not be able to purchase products from related product categories.

### Simple Example

In this example we will not allow members of the "Premium" group to purchase products that are in the "Standard" product category.

This is how we set up our filter based on these requirements. You can place this in your theme's `functions.php` file.

```
add_filter( 'groups_wc_not_purchasable_category_to_group', 'my_simple_category_to_group_filter' );
function my_simple_category_to_group_filter( $category_to_group ) {
	return array(
		'Standard' => 'Premium'
	);
}
```

The plugin will also filter the visibility of the products that are not purchasable.
If you want to change this, you can use the `groups_wc_not_purchasable_filter_visibility` filter to modify this behavior.

```
add_filter( 'groups_wc_not_purchasable_filter_visibility', 'my_groups_wc_not_purchasable_filter_visibility', 10, 3 );
function my_groups_wc_not_purchasable_filter_visibility( $filter_visibility, $visible, $product_id ) {
	return false;
}
```


### Added more entries based on the Simple Example

You could change the above and add as many entries as needed. For example, to also exclude members of the "Distributor" group from purchasing products in the "Wholesale" product category:

```
add_filter( 'groups_wc_not_purchasable_category_to_group', 'my_simple_category_to_group_filter' );
function my_simple_category_to_group_filter( $category_to_group ) {
	return array(
		'Standard'  => 'Premium',
		'Wholesale' => 'Distributor'
	);
}
```

### Advanced Example

In this example, we have three product categories that will be hidden from members of certain groups.

The product categories are:

- Basic Membership
- Premium Membership
- Platinum Membership
- Diamond Membership

The groups are:

- Basic
- Premium
- Platinum
- Diamond

We don't want to let members of the Basic, Premium, Platinum or Diamond group purchase products from product categories that belong to products that grant access to the same or inferior membership.

Members of the Platinum or Diamond group shouldn't purchase products from the inferior "Premium Membership" or "Platinum Membership" product categories.

Our Diamond members shouldn't be able to purchase products from the basic, premium, platinum or diamond product categories.

This is how we set up our filter based on these requirements. You can place this in your theme's `functions.php` file.

```
add_filter( 'groups_wc_not_purchasable_category_to_group', 'my_advanced_category_to_group_filter' );
function my_advanced_category_to_group_filter( $category_to_group ) {
	return array(
		'Basic Membership'    => 'Basic',
		'Basic Membership'    => 'Premium',
		'Basic Membership'    => 'Platinum',
		'Basic Membership'    => 'Diamond',
		'Premium Membership'  => 'Premium',
		'Premium Membership'  => 'Platinum',
		'Premium Membership'  => 'Diamond',
		'Platinum Membership' => 'Platinum'
		'Platinum Membership' => 'Diamond',
		'Diamond Membership'  => 'Diamond'
	);
}
```


changelog.txt
Groups WC Not Purchasable               Changelog              www.itthinx.com

= 1.2.0 =
* WordPress 6.4 compatible.
* Requires at least WordPress 6.0.
* WooCommerce 8.3. compatible.
* WooCommerce HPOS compatible.
* Requires at least WooCommerce 7.9.
* Fixes missing visibility filter.

= 1.1.0 =
* Added readme.txt.
* Added ABSPATH check.
* Added changelog.txt.
* Fixed formatting in plugin header.
* Added the groups_wc_not_purchasable_category_to_group filter.

= 1.0.0 =
* Initial release.


groups-wc-not-purchasable.php
<?php
/**
 * groups-wc-not-purchasable.php
 *
 * Copyright (c) www.itthinx.com
 *
 * This code is released under the GNU General Public License.
 * See COPYRIGHT.txt and LICENSE.txt.
 *
 * This code is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * This header and all notices must be kept intact.
 *
 * @author itthinx
 *
 * @since 1.0.0
 *
 * Plugin Name: Groups WooCommerce Not Purchasable
 * Plugin URI: http://www.itthinx.com/plugins/groups
 * Description: Use the Groups plugin with WooCommerce to have products that can not be purchased or viewed by group members.
 * Version: 1.2.0
 * Author: itthinx
 * WC requires at least: 7.9
 * WC tested up to: 8.3
 * Author URI: https://www.itthinx.com
 * Donate-Link: https://www.itthinx.com
 * License: GPLv3
 */

if ( !defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Restrict products that belong to a certain category to be unavailable for group members.
 */
class Groups_WC_Not_Purchasable {

	/**
	 * Maps product categories to groups:
	 *
	 * For example, products in the "Standard" product category
	 * are not available to members of the "Premium" group - use the
	 * groups_wc_not_purchasable_category_to_group filter to return
	 * and array holding:
	 *
	 * <code>array( 'Standard' => 'Premium' )</code>
	 * 
	 * Alternatively, you could fork this plugin and directly define it here.
	 */
	private static $category_to_group = array();

	/**
	 * Registers the restrictions filter.
	 */
	public static function init() {
		add_filter( 'woocommerce_is_purchasable', array( __CLASS__, 'woocommerce_is_purchasable' ), 10, 2 );
		add_filter( 'woocommerce_product_is_visible', array( __CLASS__, 'woocommerce_product_is_visible' ), 10, 2 );
		add_action( 'before_woocommerce_init', array( __CLASS__, 'before_woocommerce_init' ) );
	}

	/**
	 * Declare HPOS compatibility
	 *
	 * @since 1.2.0
	 */
	public static function before_woocommerce_init() {
		if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
		}
	}

	/**
	 * If the current user belongs to any of the groups in self::$category_to_group and the
	 * product to the corresponding category, then the product will not be purchasable by
	 * the user.
	 * 
	 * Otherwise it will return the unmodified value of $purchasable.
	 * 
	 * @param boolean $purchasable
	 * @param WC_Product|int $product
	 * 
	 * @return boolean
	 */
	public static function woocommerce_is_purchasable( $purchasable, $product ) {
		$result = $purchasable;
		if ( is_object( $product ) ) {
			$product_id = method_exists( $product, 'get_id' ) ? $product->get_id() : $product->id;
		} else {
			$product_id = intval( $product );
		}
		if ( class_exists( 'Groups_User' ) && class_exists( 'Groups_Group' ) ) {
			if ( is_user_logged_in() ) {
				$user_id = get_current_user_id();
				$user = new Groups_User( $user_id );
				$category_to_group = apply_filters(
					'groups_wc_not_purchasable_category_to_group',
					self::$category_to_group
				);
				if ( is_array( $category_to_group ) ) {
					foreach( $category_to_group as $category => $group ) {
						if ( $group = Groups_Group::read_by_name( $group ) ) {
							if ( $user->is_member( $group->group_id ) ) {
								if ( has_term( $category, 'product_cat', $product_id ) ) {
									$result = false;
									break;
								}
							}
						}
					}
				}
			}
		}
		return $result;
	}

	/**
	 * Filters the product visibility.
	 *
	 * @param boolean $visible
	 * @param int $product_id
	 *
	 * @return boolean true if product is visible
	 */
	public static function woocommerce_product_is_visible( $visible, $product_id ) {
		if ( apply_filters( 'groups_wc_not_purchasable_filter_visibility', true, $visible, $product_id ) ) {
			$visible = self::woocommerce_is_purchasable( $visible, $product_id );
		}
		return $visible;
	}
}
Groups_WC_Not_Purchasable::init();
